src/Billing/BillingProcessor/BillingLogger.php

```diff
@@ -0,0 +1,139 @@
+ * @author    Ken Chapple <ken@mi-squared.com>
+ * @author    Daniel Pflieger <daniel@growlingflea.com>
+ * @author    Terry Hill <terry@lilysystems.com>
+ * @author    Jerry Padgett <sjpadgett@gmail.com>
+ * @author    Stephen Waite <stephen.waite@cmsvt.com>
+ * @copyright Copyright (c) 2021 Ken Chapple <ken@mi-squared.com>
+ * @copyright Copyright (c) 2021 Daniel Pflieger <daniel@growlingflea.com>
+ * @copyright Copyright (c) 2014-2020 Brady Miller <brady.g.miller@gmail.com>
+ * @copyright Copyright (c) 2016 Terry Hill <terry@lillysystems.com>
+ * @copyright Copyright (c) 2017-2020 Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2018-2020 Stephen Waite <stephen.waite@cmsvt.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Billing\BillingProcessor;
+
+use OpenEMR\Common\Crypto\CryptoGen;
+
+class BillingLogger
+{
+    /**
+     * Contains an array of status messages that accumulate
+     * through the billing process
+     *
+     * @var array
+     */
+    protected $bill_info = [];
+
+    /**
+     * Contains a string that represents the results from formatting
+     * x-12 claims. This is what you see when you click the 'Logs' button
+     * on the result modal.
+     *
+     * @var false|string
+     */
+    protected $hlog;
+
+    /**
+     * @var callable
+     */
+    protected $onLogCompleteCallback;
+
+    /**
+     * Show/Hide the 'Close' button that is printed in billing_process.php
+     *
+     * @var bool
+     */
+    protected $showCloseButton = true;
+
+    public function __construct()
+    {
+        if ($GLOBALS['billing_log_option'] == 1) {
+            // Set up crypto object
+            $cryptoGen = new CryptoGen();
+
+            if (file_exists($GLOBALS['OE_SITE_DIR'] . "/documents/edi/process_bills.log")) {
+                $this->hlog = file_get_contents($GLOBALS['OE_SITE_DIR'] . "/documents/edi/process_bills.log");
+            }
+            if ($cryptoGen->cryptCheckStandard($this->hlog)) {
+                $this->hlog = $cryptoGen->decryptStandard($this->hlog, null, 'database');
+            }
+        } else { // ($GLOBALS['billing_log_option'] == 2)
+            $this->hlog = '';
+        }
+    }
+
+    public function setLogCompleteCallback(callable $onLogCompleteCallback)
+    {
+        $this->onLogCompleteCallback = $onLogCompleteCallback;
+    }
+
+    /**
+     * Called when log is done writing
+     *
+     * @return false|mixed
+     */
+    public function onLogComplete()
+    {
+        if (isset($this->onLogCompleteCallback)) {
+            call_user_func($this->onLogCompleteCallback);
+        }
+
+        return false;
+    }
+
+    public function printToScreen($message)
+    {
+        $this->bill_info[]= $message;
+    }
+
+    public function bill_info()
+    {
+        return $this->bill_info;
+    }
+
+    public function appendToLog($message)
+    {
+        $this->hlog .= $message;
```

Comment:
does this log entry get saved back to the log file?

Reply:
This is for the X12 mumbo-jumbo that gets printed to the screen when choosing "validate"